Source/Methods/Envelope Tracking/EnvelopeTracking.cs

```diff
@@ -0,0 +1,309 @@
+            scale *= ComplexSingle.FromSingle((float)Math.Pow(10.0, (dutAverageInputPower - 10.0) / 20.0)); // cascade scale values so we only have to run through the array once
+            for (int i = 0; i < iq.Length; i++)
+                iq[i] *= scale;
+
+            // get 1 ohm power trace in watts of scaled iq data
+            float[] powerTrace_W = new float[iq.Length];
+            for (int i = 0; i < iq.Length; i++)
+                powerTrace_W[i] = iq[i].Real * iq[i].Real + iq[i].Imaginary * iq[i].Imaginary;
+
+            // get lookup table input power trace in 1ohm watts
+            float[] lutDutInputPowerWattOneOhm = new float[lookUpTable.DutInputPower_dBm.Length];
+            for (int i = 0; i < lookUpTable.DutInputPower_dBm.Length; i++)
+                lutDutInputPowerWattOneOhm[i] = (float)Math.Pow(10.0, (lookUpTable.DutInputPower_dBm[i] - 10.0) / 10.0); // V^2 = 10^((Pin - 10.0)/10)
+
+            // run the trace through 1D interpolation
+            float[] rawEnvelope = LinearInterpolation1D(lutDutInputPowerWattOneOhm, lookUpTable.SupplyVoltage_V, powerTrace_W);
+
+            // create waveform to return to the user
+            Waveform envelopeWaveform = new Waveform()
+            {
+                Name = referenceWaveform.Name + "Envelope",
+                Data = referenceWaveform.Data.Clone(),
+                SignalBandwidth_Hz = 0.8 * referenceWaveform.SampleRate,
+                PAPR_dB = double.NaN, // unnecessary to calculate as it won't be used
+                BurstLength_s = double.NaN, // not applicable in ET
+                SampleRate = referenceWaveform.SampleRate,
+                // burst start and stop locations will be left as null since they also will not be used
+                IdleDurationPresent = referenceWaveform.IdleDurationPresent, // can't null a bool, so copy from reference waveform
+                RuntimeScaling = 10 * Math.Log10(0.9), // applies 10% headroom to the waveform at runtime
+                Script = referenceWaveform.Script // we will copy the script but call replace on it in the following lines to change the waveform name
+            };
+            envelopeWaveform.Script = envelopeWaveform.Script.Replace(referenceWaveform.Name, envelopeWaveform.Name); // this is better here since now we only have to add the suffix once
+            WritableBuffer<ComplexSingle> envWfmWriteBuffer = envelopeWaveform.Data.GetWritableBuffer();
+
+            // copy raw envelope data into cloned envelope waveform
+            for (int i = 0; i < rawEnvelope.Length; i++)
+                envWfmWriteBuffer[i] = ComplexSingle.FromSingle(rawEnvelope[i]);
+
+            return envelopeWaveform;
+        }
+        #endregion
+
+        #region Instrument Configuration
+        public static void ConfigureEnvelopeGenerator(NIRfsg envVsg, EnvelopeGeneratorConfiguration envVsgConfig, TrackerConfiguration trackerConfig)
+        {

```

Comment:
How do we configure instrument parameters such as reference clock? Should they be part of the lib?

Reply:
I envisioned we would do that with the SG module. Thoughts?

Reply:
I agree with @mkrmasch on this one. I don't think that we can configure this one using the SG module, because the instrument configuration function would fail. I think we ought to configure the clock and trigger export in this function. Here's what I have in LabVIEW at the moment: 
![image](https://user-images.githubusercontent.com/47825680/68975972-520b3880-07ba-11ea-9328-f2d641325f7f.png)

Reply:
why would it fail?

Reply:
Instr config takes in stuff like frequency, LO offset, etc. that doesn't make sense for baseband 5820. Also, trigger export doesn't happen except in "Configure Cont. Generation" and we are using the built-in script.

Reply:
i added it back in, but only put in reference clock source. I don't find export terminal nor hard setting the generation type as script to be necessary since they would be handled with calls to ConfigureContinuousGeneration